File: scripts/parse_exercises.py

import os
import json
import re
from typing import List, Dict, Any

# These mappings are based on the seed.ts file to ensure consistency
def map_difficulty(diff: str) -> str:
    diff_map = {
        'beginner': 'BEGINNER',
        'intermediate': 'INTERMEDIATE',
        'advanced': 'ADVANCED',
        'expert': 'EXPERT',
    }
    return diff_map.get(diff.lower(), 'BEGINNER')

def map_muscle(muscle: str) -> str:
    muscle_map = {
        'chest': 'CHEST', 'back': 'BACK', 'shoulders': 'SHOULDERS', 'biceps': 'BICEPS',
        'triceps': 'TRICEPS', 'forearms': 'FOREARMS', 'abs': 'ABS', 'core': 'ABS',
        'obliques': 'OBLIQUES', 'lower back': 'LOWER_BACK', 'lower_back': 'LOWER_BACK',
        'glutes': 'GLUTES', 'quadriceps': 'QUADRICEPS', 'quads': 'QUADRICEPS',
        'hamstrings': 'HAMSTRINGS', 'calves': 'CALVES', 'full body': 'FULL_BODY',
        'full_body': 'FULL_BODY', 'cardio': 'CARDIO', 'rear delts': 'SHOULDERS',
        'rhomboids': 'BACK', 'hip flexors': 'LEGS', 'power': 'FULL_BODY', 'grip': 'FOREARMS',
        'triceps': 'TRICEPS', 'wrist': 'FOREARMS', 'mid back': 'BACK',
        'lats': 'BACK', 'lats (width)': 'BACK', 'lats (lower)': 'BACK',
        'lats (unilateral)': 'BACK', 'lats (isolation)': 'BACK', 'lats, rhomboids': 'BACK',
        'lats, core': 'BACK', 'lats, biceps': 'BACK', 'rear delts, lats': 'BACK',
        'lats, mid back': 'BACK', 'lats, rear delts': 'BACK', 'lats (upper)': 'BACK',
        'lats (lower)': 'BACK', 'lats, chest': 'BACK', 'lats, shoulders': 'BACK'
    }
    return muscle_map.get(muscle.lower(), 'FULL_BODY')


def map_equipment(equip: str) -> str:
    equip_map = {
        'none': 'NONE', 'bodyweight': 'BODYWEIGHT', 'body weight': 'BODYWEIGHT',
        'dumbbells': 'DUMBBELLS', 'dumbbell': 'DUMBBELLS', 'db': 'DUMBBELLS',
        'db + bench': 'DUMBBELLS', 'heavy db': 'DUMBBELLS', 'light dbs': 'DUMBBELLS',
        'barbell': 'BARBELL', 'bar': 'BARBELL', 'barbell + bench': 'BARBELL',
        'bar/rings': 'BARBELL', 'bar/trx': 'BARBELL', 'bar + band': 'BARBELL',
        'bar + weight': 'BARBELL',
        'kettlebell': 'KETTLEBELL', 'cables': 'CABLES', 'cable': 'CABLES',
        'high pulley': 'CABLES', 'low pulley': 'CABLES',
        'machines': 'MACHINES', 'machine': 'MACHINES', 'plate-loaded': 'MACHINES',
        'landmine': 'MACHINES', 'machine lat pulldown': 'MACHINES',
        'resistance bands': 'RESISTANCE_BANDS', 'bands': 'RESISTANCE_BANDS', 'band': 'RESISTANCE_BANDS',
        'band + anchor': 'RESISTANCE_BANDS',
        'trx': 'TRX', 'suspension': 'TRX', 'rings': 'TRX',
        'pull-up bar': 'PULL_UP_BAR', 'pull up bar': 'PULL_UP_BAR', 'pullup bar': 'PULL_UP_BAR',
        'bench': 'BENCH', 'incline': 'BENCH', 'decline': 'BENCH', 'db + incline': 'BENCH',
        'stability ball': 'STABILITY_BALL', 'swiss ball': 'STABILITY_BALL',
        'foam roller': 'FOAM_ROLLER', 'jump rope': 'JUMP_ROPE', 'treadmill': 'TREADMILL',
        'bike': 'BIKE', 'rowing': 'ROWING',
    }
    # First, check for exact match
    if equip.lower() in equip_map:
        return equip_map[equip.lower()]
    # If no exact match, check for keywords
    for keyword, eq_type in equip_map.items():
        if keyword in equip.lower():
            return eq_type
    return 'NONE'

def map_category(cat: str) -> str:
    # We will derive category from the file path for now
    return cat.upper()


def parse_markdown_table(content: str, muscle_group_from_path: str) -> List[Dict[str, Any]]:
    """Parses all markdown tables in a string and returns a list of exercise dicts."""
    
    # Regex to find markdown tables
    table_regex = re.compile(
        r"\|(?P<header>.+?)\|\n"  # Header row
        r"\|(?P<separator>[-|: ]+)\|\n"  # Separator row
        r"(?P<rows>(?:\|.*?(?:\n|$))+)",  # Data rows
        re.MULTILINE
    )
    
    exercises = []
    
    for match in table_regex.finditer(content):
        header = [h.strip().lower() for h in match.group('header').split('|') if h.strip()]
        
        raw_rows = match.group('rows').strip().split('\n')
        
        for row_str in raw_rows:
            if not row_str.strip().startswith('|'):
                continue

            values = [v.strip() for v in row_str.split('|') if v.strip()]
            
            if len(values) != len(header):
                # Malformed row, skip
                continue

            row_data = dict(zip(header, values))

            # --- Data Transformation ---
            secondary_muscles_raw = row_data.get('secondary', '')
            secondary_muscles_list = [m.strip() for m in re.split(r',|/', secondary_muscles_raw) if m.strip()]

            equipment_raw = row_data.get('equipment', 'bodyweight')
            equipment_list = [e.strip() for e in re.split(r',|/', equipment_raw) if e.strip()]

            # This structure mirrors the `ExerciseData` interface in seed.ts
            exercise = {
                "id": row_data.get('id', '').strip(),
                "name_en": row_data.get('exercise', 'Unknown Exercise').strip(),
                "name_ar": f"{row_data.get('exercise', 'Unknown').strip()} (AR)", # Placeholder
                "description_en": f"A {muscle_group_from_path} exercise focusing on the {row_data.get('primary', 'muscles')}.", # Placeholder
                "description_ar": f"تمرين {muscle_group_from_path} يركز على {row_data.get('primary', 'muscles')}.", # Placeholder
                "category": muscle_group_from_path.upper(),
                "primary_muscle": map_muscle(row_data.get('primary', muscle_group_from_path)),
                "secondary_muscles": [map_muscle(m) for m in secondary_muscles_list],
                "equipment": [map_equipment(e) for e in equipment_list],
                "difficulty": map_difficulty(row_data.get('difficulty', 'beginner')),
                "instructions_en": [ # Placeholder
                    "Step 1: Assume starting position.",
                    "Step 2: Perform the movement.",
                    "Step 3: Return to start."
                ],
                "instructions_ar": [
                    "الخطوة ١: اتخذ وضعية البداية.",
                    "الخطوة ٢: قم بأداء الحركة.",
                    "الخطوة ٣: عد إلى وضعية البداية."
                ],
                "tips_en": [],
                "tips_ar": [],
                "is_time_based": False,
                "default_sets": 3,
                "default_reps": 12,
                "default_duration": None,
                "default_rest": 60,
                "tags": [muscle_group_from_path] + [e.lower().replace(' ', '-') for e in equipment_list]
            }
            exercises.append(exercise)
            
    return exercises


def main():
    """Main function to find MD files, parse them, and write JSON."""
    
    root_docs_dir = os.path.abspath(os.path.join(__file__, '..', '..', 'docs'))
    exercises_dir = os.path.join(root_docs_dir, 'exercises')
    
    if not os.path.isdir(exercises_dir):
        print(f"Error: Directory not found '{exercises_dir}'")
        return

    print(f"Starting scan in: {exercises_dir}")
    total_files_processed = 0
    total_exercises_found = 0

    for root, _, files in os.walk(exercises_dir):
        for file in files:
            if file.endswith('.md'):
                md_path = os.path.join(root, file)
                print(f"\nProcessing: {md_path}")
                
                # Derive muscle group from the parent directory name
                muscle_group = os.path.basename(root)

                try:
                    with open(md_path, 'r', encoding='utf-8') as f:
                        content = f.read()
                    
                    exercises = parse_markdown_table(content, muscle_group)

                    if not exercises:
                        print("  - No tables found or parsed.")
                        continue
                    
                    total_exercises_found += len(exercises)
                    total_files_processed += 1
                    
                    # Write to JSON file
                    json_path = os.path.splitext(md_path)[0] + '.json'
                    with open(json_path, 'w', encoding='utf-8') as f:
                        json.dump(exercises, f, indent=2, ensure_ascii=False)
                    
                    print(f"  [OK] Found {len(exercises)} exercises. Saved to {os.path.basename(json_path)}")

                except Exception as e:
                    print(f"  [ERROR] Error processing file {file}: {e}")

    print("\n-----------------------------------------")
    print("All files processed!")
    print(f"Total Markdown files parsed: {total_files_processed}")
    print(f"Total exercises generated: {total_exercises_found}")
    print("-----------------------------------------\\n")


if __name__ == '__main__':
    main()

File: apps/api/prisma/parse-exercises.ts
import * as fs from 'fs';
import * as path from 'path';

interface Exercise {
  id: string;
  name_en: string;
  name_ar: string;
  description_en: string;
  description_ar: string;
  category: string;
  primary_muscle: string;
  secondary_muscles: string[];
  equipment: string[];
  difficulty: string;
  instructions_en: string[];
  instructions_ar: string[];
  tips_en: string[];
  tips_ar: string[];
  is_time_based: boolean;
  default_sets: number;
  default_reps: number;
  default_duration: number | null;
  default_rest: number;
  tags: string[];
}

function parseExerciseMD(content: string, category: string, muscle: string): Exercise[] {
  const exercises: Exercise[] = [];

  // Match exercise headers like: ### GL001 - Barbell Hip Thrust
  const exercisePattern = /###\s+([A-Z]{1,3}\d{3})\s+-\s+(.+?)(?=\n)/g;
  const tablePattern = /\|\s*\*\*([^*]+)\*\*\s*\|\s*([^|]+)\s*\|/g;

  let match;
  const exerciseBlocks: { id: string; name: string; startIndex: number; endIndex: number }[] = [];

  // Find all exercise headers
  while ((match = exercisePattern.exec(content)) !== null) {
    exerciseBlocks.push({
      id: match[1],
      name: match[2].trim(),
      startIndex: match.index,
      endIndex: content.length
    });
  }

  // Set end indices
  for (let i = 0; i < exerciseBlocks.length - 1; i++) {
    exerciseBlocks[i].endIndex = exerciseBlocks[i + 1].startIndex;
  }

  // Parse each exercise block
  for (const block of exerciseBlocks) {
    const blockContent = content.substring(block.startIndex, block.endIndex);

    // Extract attributes from table
    const attributes: Record<string, string> = {};
    let tableMatch;
    while ((tableMatch = tablePattern.exec(blockContent)) !== null) {
      attributes[tableMatch[1].toLowerCase()] = tableMatch[2].trim();
    }

    // Extract form cues/instructions
    const instructionsMatch = blockContent.match(/\*\*Form Cues:\*\*\n([\s\S]*?)(?=\n\n|\*\*Common|$)/);
    const instructions: string[] = [];
    if (instructionsMatch) {
      const lines = instructionsMatch[1].split('\n');
      for (const line of lines) {
        const cleaned = line.replace(/^-\s*/, '').trim();
        if (cleaned) instructions.push(cleaned);
      }
    }

    // Extract tips from common mistakes
    const tipsMatch = blockContent.match(/\*\*Common Mistakes:\*\*\n([\s\S]*?)(?=\n###|$)/);
    const tips: string[] = [];
    if (tipsMatch) {
      const lines = tipsMatch[1].split('\n');
      for (const line of lines) {
        const cleaned = line.replace(/^-\s*/, '').trim();
        if (cleaned) tips.push(cleaned);
      }
    }

    // Map difficulty
    const difficultyMap: Record<string, string> = {
      'beginner': 'BEGINNER',
      'intermediate': 'INTERMEDIATE',
      'advanced': 'ADVANCED',
      'expert': 'EXPERT'
    };

    // Map equipment
    const equipmentMap: Record<string, string> = {
      'barbell': 'BARBELL',
      'dumbbell': 'DUMBBELLS',
      'dumbbells': 'DUMBBELLS',
      'bench': 'BENCH',
      'cable': 'CABLES',
      'cables': 'CABLES',
      'machine': 'MACHINES',
      'bodyweight': 'BODYWEIGHT',
      'kettlebell': 'KETTLEBELL',
      'bands': 'RESISTANCE_BANDS',
      'resistance bands': 'RESISTANCE_BANDS',
      'pull-up bar': 'PULL_UP_BAR',
      'trx': 'TRX',
      'none': 'NONE'
    };

    // Map muscle groups
    const muscleMap: Record<string, string> = {
      'gluteus maximus': 'GLUTES',
      'glutes': 'GLUTES',
      'hamstrings': 'HAMSTRINGS',
      'quadriceps': 'QUADRICEPS',
      'quads': 'QUADRICEPS',
      'calves': 'CALVES',
      'hip flexors': 'QUADRICEPS',
      'core': 'ABS',
      'abs': 'ABS',
      'obliques': 'OBLIQUES',
      'chest': 'CHEST',
      'back': 'BACK',
      'lats': 'BACK',
      'shoulders': 'SHOULDERS',
      'biceps': 'BICEPS',
      'triceps': 'TRICEPS',
      'forearms': 'FOREARMS',
      'lower back': 'LOWER_BACK'
    };

    // Parse equipment string
    const equipmentStr = attributes['equipment'] || 'bodyweight';
    const equipment = equipmentStr.split(',').map(e => {
      const cleaned = e.trim().toLowerCase();
      return equipmentMap[cleaned] || 'NONE';
    });

    // Parse primary muscle
    const primaryStr = (attributes['primary'] || muscle).toLowerCase();
    const primaryMuscle = muscleMap[primaryStr] || 'FULL_BODY';

    // Parse secondary muscles
    const secondaryStr = attributes['secondary'] || '';
    const secondaryMuscles = secondaryStr.split(',')
      .map(m => muscleMap[m.trim().toLowerCase()])
      .filter(Boolean);

    // Parse difficulty
    const difficultyStr = (attributes['difficulty'] || 'beginner').toLowerCase();
    const difficulty = difficultyMap[difficultyStr] || 'BEGINNER';

    // Create Arabic name (placeholder - would need translation)
    const nameAr = block.name + ' (AR)';

    // Create description
    const descriptionEn = `A ${category.toLowerCase()} exercise focusing on the ${primaryStr}.`;
    const descriptionAr = `تمرين ${category.toLowerCase()} يركز على ${primaryStr}.`;

    exercises.push({
      id: block.id,
      name_en: block.name,
      name_ar: nameAr,
      description_en: descriptionEn,
      description_ar: descriptionAr,
      category: category.toUpperCase(),
      primary_muscle: primaryMuscle,
      secondary_muscles: secondaryMuscles,
      equipment: equipment,
      difficulty: difficulty,
      instructions_en: instructions.length > 0 ? instructions : [
        'Step 1: Assume starting position.',
        'Step 2: Perform the movement.',
        'Step 3: Return to start.'
      ],
      instructions_ar: [
        'الخطوة ١: اتخذ وضعية البداية.',
        'الخطوة ٢: قم بأداء الحركة.',
        'الخطوة ٣: عد إلى وضعية البداية.'
      ],
      tips_en: tips,
      tips_ar: [],
      is_time_based: false,
      default_sets: 3,
      default_reps: 12,
      default_duration: null,
      default_rest: 60,
      tags: [category.toLowerCase(), muscle.toLowerCase()]
    });
  }

  return exercises;
}

function findMdFiles(dir: string): string[] {
  const files: string[] = [];
  if (!fs.existsSync(dir)) return files;

  const items = fs.readdirSync(dir);
  for (const item of items) {
    const fullPath = path.join(dir, item);
    const stat = fs.statSync(fullPath);
    if (stat.isDirectory()) {
      files.push(...findMdFiles(fullPath));
    } else if (item.endsWith('.md')) {
      files.push(fullPath);
    }
  }
  return files;
}

async function main() {
  console.log('\n╔═══════════════════════════════════════════════════════╗');
  console.log('║   📝 PARSING EXERCISE MARKDOWN FILES                   ║');
  console.log('╚═══════════════════════════════════════════════════════╝\n');

  const docsDir = path.join(__dirname, '../../../docs/exercises');
  const mdFiles = findMdFiles(docsDir);

  console.log(`📂 Found ${mdFiles.length} markdown files\n`);

  let totalExercises = 0;

  for (const filePath of mdFiles) {
    const relativePath = path.relative(docsDir, filePath);
    const category = path.dirname(relativePath);
    const muscle = path.basename(filePath, '.md');

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const exercises = parseExerciseMD(content, category, muscle);

      if (exercises.length > 0) {
        // Write JSON file
        const jsonPath = filePath.replace('.md', '.json');
        fs.writeFileSync(jsonPath, JSON.stringify(exercises, null, 2));
        console.log(`  ✅ ${relativePath}: ${exercises.length} exercises`);
        totalExercises += exercises.length;
      } else {
        console.log(`  ⚠️ ${relativePath}: No exercises found`);
      }
    } catch (error: any) {
      console.error(`  ❌ ${relativePath}: ${error.message}`);
    }
  }

  console.log('\n═══════════════════════════════════════════════════════');
  console.log(`📊 TOTAL EXERCISES PARSED: ${totalExercises}`);
  console.log('═══════════════════════════════════════════════════════\n');
}

main().catch(console.error);
